deps.ml

```diff
@@ -0,0 +1,185 @@
+
+let describe_external_lib_deps =
+  Lazy.from_fun (fun _ ->
+      sexp @@ dune_describe_external_lib_deps ()
+      |> Describe_external_lib.describe_extern_of_sexp)
+
+let describe_bin_of_entries =
+  Lazy.from_fun (fun _ ->
+      sexp @@ dune_describe_entries ()
+      |> Describe_entries.entries_of_sexp
+      |> Describe_entries.items_bin_of_entries)
+
+let has_dune_subproject = function
+  | "." | "" -> false
+  | dir -> Sys.file_exists (Filename.concat dir "dune-project")
+
+let rec should_use_dir ~dir_types path =
+  match Hashtbl.find_opt dir_types path with
+  | Some x -> x
+  | None ->
+    let r =
+      match Astring.String.cut ~sep:"/" ~rev:true path with
+      | Some (parent, _) ->
+        if should_use_dir ~dir_types parent then (
+          not (has_dune_subproject path)
+        ) else false
+      | None ->
+        not (has_dune_subproject path)
+    in
+    Hashtbl.add dir_types path r;
+    r
+
+let copy_rules () =
+  Lazy.force describe_external_lib_deps
+  |> List.concat_map
+    (fun d_item ->
+       d_item
+       |> Describe_external_lib.get_item
+       |> (fun (item:Describe_external_lib.item) -> String.cat item.source_dir "/dune")
+       |> (Dune_rules.Copy_rules.get_copy_rules))
+  |> Dune_rules.Copy_rules.copy_rules_map
+
+let bin_of_entries () = Lazy.force describe_bin_of_entries
+
+let is_bin_name_of_describe_lib bin_name (item:Describe_external_lib.item) =
+  item.extensions
+  |> List.exists (fun extension ->
+      String.equal bin_name (String.cat item.name extension))
+
+let find_package_of_exe (item:Describe_external_lib.item) =
+  match item.package with
+  | Some p -> Some p
+  | None ->
+    (* Only allow for private executables to find the package *)
+    item.extensions
+    |> List.find_map (fun extension ->
+        Option.map
+          (fun bin_name ->
+             Option.map
+               (fun (item:Describe_entries.item) -> item.package) (Item_map.find_opt bin_name @@ bin_of_entries ()))
+          (Dune_rules.Copy_rules.find_dest_name ~name:(String.cat item.name extension) @@ copy_rules ()))
+    |> Option.join
+
+let resolve_internal_deps d_items items_pkg =
+  (* After the d_items are filtered to the corresponding package request,
+   * we need to include the internal_deps in order to reach all the deps.
+   * If the internal dep is a public library we skip the recursive resolve
+   * because it will be resolve with separate request *)
+  let open Describe_external_lib in
+  let get_name = function
+    | Lib item  -> String.cat item.name ".lib"
+    | Exe item  -> String.cat item.name ".exe"
+    | Test item -> String.cat item.name ".test"
+  in
+  let d_items_lib =
+    d_items
+    |> List.filter_map (fun d_item ->
+        match is_lib_item d_item with
+        | true ->
+          d_item
+          |> get_item
+          |> (fun (item:Describe_external_lib.item) ->
+              (String.cat item.name ".lib", Lib item))
+          |> Option.some
+        | false -> None)
```

Comment:
Prefer using `if...then...else` rather than `match` expressions on booleans. The rest of the code can be made clearer like this, imo.

```suggestion
        if is_lib_item d_item then
          let item = get_item d_item in
          Some (item.Describe_external_lib.name ^ ".lib", Lib item)
        else None)
```